File: ocaml/pr020.ml
(*
 * Problem 20
 *
 * 21 June 2002
 *
 * n! means n × (n − 1) × ... × 3 × 2 × 1
 *
 * For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,
 * and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0
 * + 0 = 27.
 *
 * Find the sum of the digits in the number 100!
 *)

open Printf
open Num

let fact n =
  let rec loop result = function
    | 0 -> result
    | n -> loop (result */ num_of_int n) (n-1) in
  loop (num_of_int 1) n

let ten = num_of_int 10
let zero = num_of_int 0
let sum_digits n =
  let rec loop result n =
    if n =/ zero then result
    else loop (result + (int_of_num (mod_num n ten))) (quo_num n ten) in
  loop 0 n

let () = printf "%d\n" (sum_digits (fact 100))

File: fsharp/pr020.fs
// Problem 20
//
// 21 June 2002
//
//
// n! means n x (n − 1) x ... x 3 x 2 x 1
//
// For example, 10! = 10 x 9 x ... x 3 x 2 x 1 = 3628800,
// and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 =
// 27.
//
// Find the sum of the digits in the number 100!
//
// 648

let fact (n : int) =
    seq { 1I .. bigint n }
    |> Seq.reduce ( * )

let digitSum n =
    let rec loop n total =
        if n = 0I then total
        else loop (n / 10I) (total + int (n % 10I))
    loop n 0

let pr20 () = fact 100 |> digitSum

printfn "%A" (pr20 ())
